core/src/main/scala/kafka/controller/ControllerChannelManager.scala

```diff
@@ -138,9 +138,11 @@ class ControllerChannelManager(controllerContext: ControllerContext, config: Kaf
 
   private def removeExistingBroker(brokerState: ControllerBrokerStateInfo) {
     try {
+      // shutdown the RequestSendThread before closing the NetworkClient to
+      // avoid the concurrent use of the non-threadsafe classes as described in KAFKA-4959
+      brokerState.requestSendThread.shutdown()
```

Comment:
First of all, great work figuring this one out!

I think this change makes things better and may be OK to include as is since @onurkaraman is planning to change the threading model for the Controller. However, it's worth pointing out that this is still not safe. The fact that we access the `NetworkClient` from two different threads without any synchronization means that the code is basically buggy although the bugs are likely to happen rarely (particularly on x86/x86-64, which is strongly ordered) and they will be even harder to debug. So, it would be good to mention this in the comment and file a JIRA as well.